### ISSUE TITLE
Inconsistent change-tracking behavior when passing all assigns

### ISSUE THREAD

**Issue by SteffenDE** [phoenixframework/phoenix_live_view#2995]
### Environment

* Elixir version (elixir -v): 1.16.0
* Phoenix version (mix deps): 1.7.10
* Phoenix LiveView version (mix deps): main

### Actual behavior

Sometimes it is convenient to pass all assigns from one function component to another. Let's look at one example:

```elixir
def my_function_component(assigns) do
  ~H"""
  <.internal_header {assigns} />
  ...
  <.internal_footer {assigns />
  """
end

defp internal_header(assigns) do
  ...
end

...
```

When doing this, the diff over the wire explodes, as any change always sends the all data for `internal_header` and `internal_footer`. As these are all function components it should be possible to optimize this. Indeed, this works just fine if we change the code like this:

```elixir
def my_function_component(assigns) do
  ~H"""
  <%= internal_header(assigns) %>
  ...
  <%= internal_footer(assigns) %>
  """
end
```

### Expected behavior

As these are function components and they have the special `__changed__` assign, I think that the first syntax should behave the same. I attached an example single-file script to reproduce.

```elixir
Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:jason, "~> 1.0"},
  {:phoenix, "~> 1.7.10", override: true},
  {:phoenix_live_view, github: "phoenixframework/phoenix_live_view", branch: "main"}
])

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

 def mount(_params, _session, socket) do
    socket
    |> assign(a: 1, b: 2, c: 3)
    |> then(&{:ok, &1})
  end

  def handle_event("set_c", _, socket) do
    {:noreply, assign(socket, :c, :rand.uniform(10) |> floor())}
  end

  def render("live.html", assigns) do
    ~H"""
    <script src="https://cdn.jsdelivr.net/npm/phoenix@1.7.10/priv/static/phoenix.min.js"></script>
    <script src="https://cdn.jsdelivr.net/gh/phoenixframework/phoenix_live_view@main/priv/static/phoenix_live_view.js"></script>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket, {
        hooks: {
          FakeHook: {
            mounted() {}
          }
        }
      })
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 1.1em; }
    </style>
    <%= @inner_content %>
    """
  end

  def render(assigns) do
    ~H"""
    <.my_component a={1} b={2} c={@c} />

    <button phx-click="set_c">Set C</button>
    """
  end

  defp my_component(assigns) do
    ~H"""
    <h1>Test</h1>

    <.my_other_component {assigns} />
    <%!-- <%= my_other_component(assigns) %> --%>
    """
  end

  defp my_other_component(assigns) do
    IO.inspect(assigns)

    ~H"""
    <div>
      <span>A: <%= @a %></span>
      <span>B: <%= @b %></span>
      <span>C: <%= @c %></span>
    </div>
    """
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :index)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)
  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)
```

In this case we can see that when we use `<%= my_other_component(assigns) %>` the function component logs 
```elixir
%{c: 8, a: 1, b: 2, __changed__: %{c: true}}
```
and only the diff for `c` is sent. If we change it to `<.my_other_component {assigns} />` we get
```elixir
%{
  c: 8,
  a: 1,
  b: 2,
  __changed__: %{c: true, a: true, b: true, __changed__: true}
}
```
instead and the a, b and c all get sent over the wire.

I've read in some blog posts about change-tracking pitfalls that passing all assigns disables change-tracking, which explains this, but I did not find any information why this happens and it does not seem to be necessary here.

Thank you!

**Comment by josevalim** [comment]
I am not sure if we can/should fix this. The short answer is that

    <.internal_header {assigns} />

is different from

    <%= internal_header(assigns) %>

in how they are interpreted. The former would be equivalent to this:

    <%= internal_header %{a: assigns.a, b: assigns.b, c: assigns.c} %>

And that's because the interpolation syntax is meant to pass _any_ attribute, we cannot just assume a particular meaning.

What we could do is to emit a warning if `<.internal_header {assigns} />` is found but passing `assigns` around is already discouraged anyway. Thoughts?

**Comment by SteffenDE** [comment]
I think adding a note to the documentation somewhere (https://hexdocs.pm/phoenix_live_view/assigns-eex.html#pitfalls maybe?) would be sufficient. A warning would be even better though, because when refactoring a large component "internally" I've seen this multiple times in our codebase. Maybe even mentioning that `<%= component(assigns) %>` should be used instead if that is the "correct" way to do this.

> passing assigns around is already discouraged anyway

I am aware of this, but I'm not sure where I got this from. Do you know where this is mentioned in the documentation?

So I'm fine with either solution, I just think that currently it is quite easy to "do it wrong" and only notice it because the diff messages get really large.

> And that's because the interpolation syntax is meant to pass any attribute, we cannot just assume a particular meaning.

What if it was equivalent to:

```elixir
<%= internal_header %{a: assigns.a, b: assigns.b, c: assigns.c, __changed__: assigns.__changed__} %>
```

**Comment by josevalim** [comment]
Can you please send a PR to the pitfalls page? Mention that you want to discourage passing `assigns` between components. If you have to do it, use the regular call syntax.

> What if it was equivalent to:

The problem is in detecting and assuming that the field you are interpolating to the component has a `__changed__` field and that the changes are correct (especially if you pass other attributes to the component). It only works in the case where you pass `assigns`. We could special case it, but I am unsure. :S

**Comment by SteffenDE** [comment]
Closed via #2998.
